sources/web/datalab/polymer/components/item-list/item-list.ts

```diff
@@ -251,12 +257,60 @@ class ItemListElement extends Polymer.Element {
       const shadow = '0px ' + yOffset + 'px 10px -5px #ccc';
       headerContainer.style.boxShadow = shadow;
     });
+
+    // Initial sort
+    this._sortBy(0);
   }
 
   resetFilter() {
     this._filterString = '';
   }
 
+  _columnButtonClicked(e: any) {
+    this._sortBy(e.model.itemsIndex);
+  }
+
+  _sortBy(column: number) {
+    // If sort is requested on the current sort column, reverse the sort order.
+    // Otherwise, set the current sort column to that.
+    if (this._currentSort.column === column) {
+      this._currentSort.asc = !this._currentSort.asc;
+    } else {
+      this._currentSort = {
+        asc: true,
+        column,
+      };
+    }
+
+    this.$.list.sort = (a: ItemListRow, b: ItemListRow) => {
+      // Bail out of sort if no columns have been set yet.
+      if (!this.columns.length) {
+        return;
+      }
+      let compResult = -1;
+      if ((this.columns[column].type === 'string' &&
+          a.columns[column].toLowerCase() > b.columns[column].toLowerCase()) ||
+          (this.columns[column].type === 'date' &&
+          new Date(a.columns[column]) > new Date(b.columns[column]))) {
```

Comment:
Creating Date objects from a string is "[strongly discouraged](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Date/parse)" due to differences in parser implementation. Plus, isn't this now a locale-dependent string?
Now that item-list knows the type of each column, you could make DatalabFile.getColumnValues return the value as a date object and do the conversion to a string when displaying the column value in item-list. That would mean that a.columns[column] at this point would be a date object, so you can just directly compare. It also would centralize the date-to-string conversion code in one place in item-list rather than in every file-manager that returns a date column.

Reply:
According to that link, this is only pre-ES5?
Aren't we doing the conversion from string to Date anyway, because we get the value as a string from the upstream service?

Reply:
The date strings coming from Drive are in ISO8601 format, which is safe to parse. As long as Americans use mm/dd/yyyy and Europeans use dd/mm/yyyy, you won't be able to unambiguously parse localized date strings without knowing _a priori_ what format they are in. Date parsing of anything other than a known fixed format is brittle.
Sorting on the actual date here rather than trying to parse the string will make it possible to display more sophisticated date strings, such as the year-or-time format of <code>ls -l</code>.

Reply:
Currently, the item-list's template prints the value of the column as specified in the `rows` array, which means a value of type `Date` will be converted to string by calling its `toString` method. We have no control over this as far as I can see. The only way out of this is by keeping two versions of the `rows` data, one with `Date` objects and one with their respective `string` versions in whichever format we want, but this complicates things quite a bit.

What if we convert all date strings into Date objects, then back to ISO-format strings before passing them to the item-list? This way the item-list itself only parses standard formats.

Reply:
I think you should be able to do something like this in item-list.html:
```
  <template is="dom-repeat" items="{{row.columns}}" as="column" indexAs="columnIndex">
    <div class="column">
      <span>[[_computeColumnDisplay(column, columnIndex)]]</span>
    </div>
  </template>
```
_computeColumnDisplay can then use this.columns[columnIndex].type to determine how to format the column value.

Reply:
Ok I modified the item list to support Date, string, and number types, and do proper sorting for these. We can think about the `ls -l` semantics separately. PTAL.